WJ-1139: add forum schema, prune references, and seed sequences

Syxless opened the pull request:
Implements the forum schema (groups, categories, threads, locks, posts, post revisions) with site-level composite FKs, actor fields, soft-deletes, discussion-thread uniqueness, and indexes for ordering/activity/lookups.
Adds forum compatibility ID bumps in the seeder.
Updates `TextService::prune()` to keep forum post revision text hashes.

Tests: ran.

Syxless committed:
WJ-1139: add forum schema, prune references, and seed sequences

emmiegit commented on deepwell/migrations/20220906103252_deepwell.sql:
```diff
@@ -732,6 +732,186 @@ CREATE TABLE filter (
+    layout TEXT,
+
+    UNIQUE (forum_group_id, sort_index),
+    UNIQUE (forum_category_id, site_id),
```
Is this for an index? I'm not sure what structural constraint this represents, given `forum_category_id` is unique by itself.

Syxless replied:
`UNIQUE (forum_category_id, site_id)` is intentional for composite FK integrity on denormalized site-scoped references, not for indexing.

emmiegit commented on deepwell/migrations/20220906103252_deepwell.sql:
```diff
@@ -732,6 +732,186 @@ CREATE TABLE filter (
+    FOREIGN KEY (forum_group_id, site_id) REFERENCES forum_group(forum_group_id, site_id)
+);
+
+-- Threads live under a category.
```
```suggestion
-- Threads live in a category (but can be moved between them).
```

emmiegit commented on deepwell/migrations/20220906103252_deepwell.sql:
```diff
@@ -732,6 +732,186 @@ CREATE TABLE filter (
+    forum_category_id BIGINT NOT NULL REFERENCES forum_category(forum_category_id),
+    forum_group_id BIGINT NOT NULL REFERENCES forum_group(forum_group_id),
+    site_id BIGINT NOT NULL REFERENCES site(site_id),
+    page_id BIGINT REFERENCES page(page_id) UNIQUE,
```
```suggestion
    page_id BIGINT REFERENCES page(page_id) UNIQUE,  -- For page discussion threads (NULL = regular thread)
```

emmiegit commented on deepwell/migrations/20220906103252_deepwell.sql:
```diff
@@ -732,6 +732,186 @@ CREATE TABLE filter (
+    latest_revision_id BIGINT,
+
+    CHECK ((deleted_by IS NULL) = (deleted_at IS NULL)),
+    UNIQUE (forum_post_id, site_id),
```
Same here.

Syxless replied:
Applied same rationale for threads, `UNIQUE (forum_thread_id, site_id)` supports composite FK checks from child tables.

emmiegit commented on deepwell/migrations/20220906103252_deepwell.sql:
```diff
@@ -732,6 +732,186 @@ CREATE TABLE filter (
+
+    CHECK ((updated_by IS NULL) = (updated_at IS NULL)),
+    CHECK ((deleted_by IS NULL) = (deleted_at IS NULL)),
+    UNIQUE (forum_thread_id, site_id),
```
Same here.

Syxless replied:
Applied same rationale for posts, `UNIQUE (forum_post_id, site_id)` supports composite FK checks from child tables

emmiegit commented on deepwell/migrations/20220906103252_deepwell.sql:
```diff
@@ -732,6 +732,186 @@ CREATE TABLE filter (
+    FOREIGN KEY (forum_group_id, site_id) REFERENCES forum_group(forum_group_id, site_id)
+);
+
+-- Locks on posts (one active lock per thread at a time).
```
```suggestion
-- Locks on posts (one active lock per post at a time).
```

emmiegit commented on deepwell/migrations/20220906103252_deepwell.sql:
```diff
@@ -732,6 +732,186 @@ CREATE TABLE filter (
+    user_id BIGINT NOT NULL REFERENCES "user"(user_id),
+    created_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
+    updated_at TIMESTAMP WITH TIME ZONE,
+    revision_number INT NOT NULL,
```
```suggestion
    revision_number INTEGER NOT NULL CHECK (revision_number >= 0),
```

emmiegit commented on deepwell/migrations/20220906103252_deepwell.sql:
```diff
@@ -732,6 +732,186 @@ CREATE TABLE filter (
+-- Latest revision FK on posts, now that the revision table exists.
+ALTER TABLE forum_post
+    ADD CONSTRAINT forum_post_latest_revision_fk
+        FOREIGN KEY (latest_revision_id) REFERENCES forum_post_revision(forum_post_revision_id);
```
👍

emmiegit commented on deepwell/src/database/seeder/mod.rs:
```diff
@@ -460,13 +460,14 @@ pub async fn seed(state: &ServerState) -> Result<()> {
     restart_sequence_with(&txn, "page_page_id_seq", 3000000000).await?;
     restart_sequence_with(&txn, "page_revision_revision_id_seq", 3000000000).await?;
     restart_sequence_with(&txn, "page_category_category_id_seq", 100000000).await?;
-
```
Note, the doc describing these is https://github.com/scpwiki/wikijump/blob/develop/docs/compatibility-ids.md

emmiegit commented on deepwell/src/database/seeder/mod.rs:
```diff
@@ -460,13 +460,14 @@ pub async fn seed(state: &ServerState) -> Result<()> {
+    restart_sequence_with(&txn, "forum_post_revision_forum_post_revision_id_seq", 3000000000)
+        .await?;
+    restart_sequence_with(&txn, "forum_thread_lock_forum_thread_lock_id_seq", 20000000).await?;
+    restart_sequence_with(&txn, "forum_post_lock_forum_post_lock_id_seq", 20000000).await?;
```
Wikidot doesn't expose IDs for forum groups and forum post revisions publicly, so we don't have compatibility IDs for those. Forum thread and post lock history is a Wikijump concept so compatibility IDs don't apply.

emmiegit commented on deepwell/src/services/text.rs:
```diff
@@ -150,8 +150,15 @@ impl TextService {
+                    ))
+                    .add(not_in_column!(
+                        Alias::new("forum_post_revision"),
+                        Alias::new("compiled_html_hash"),
```
Ah I see, temporary measure until new `src/models` files are generated. I'll do that in a separate PR because the autogen tool needs some manual correction unfortunately.

Syxless committed:
Update deepwell/migrations/20220906103252_deepwell.sql

Co-authored-by: Emmie <emmie.maeda@gmail.com>

Syxless committed:
Update deepwell/migrations/20220906103252_deepwell.sql

Co-authored-by: Emmie <emmie.maeda@gmail.com>

Syxless committed:
Update deepwell/migrations/20220906103252_deepwell.sql

Co-authored-by: Emmie <emmie.maeda@gmail.com>

Syxless committed:
Update deepwell/migrations/20220906103252_deepwell.sql

Co-authored-by: Emmie <emmie.maeda@gmail.com>

Syxless committed:
Merge branch 'develop' into fix/wj-1139-forum-schema

Syxless committed:
WJ-1139: finalize forum schema constraints and compatibility IDs

- add page->forum_thread FK for discussion threads

- lock forum_post_lock to posts instead of threads

- seed forum category/thread/post compatibility ID sequences
